src/components/editors/DictionaryEditor.tsx

```diff
@@ -16,22 +16,100 @@
 
 import { useInput } from '../../hooks/useInput';
 import ConfigEditor from './ConfigEditor';
-import { DictionaryEntity } from '../../models/Dictionary';
+import { downloadFile, isXMLValid } from '../../helpers/files';
+import { usePrevious } from '../../hooks/usePrevious';
+import { observer } from 'mobx-react-lite';
+import { createUseStyles } from 'react-jss';
+import Icon from '../Icon';
+import { buttonReset, visuallyHidden } from '../../styles/mixins';
+import { useSelectedDictionaryStore } from '../../hooks/useSelectedDictionaryStore';
 
-interface DictionaryEditorProps {
-	dictionary: DictionaryEntity | null;
-}
+const useStyle = createUseStyles({
+	controls: {
+		display: 'flex',
+		justifyContent: 'flex-end',
+		alignItems: 'center'
+	},
+	applyChanges: {
+		...buttonReset(),
+	},
+	download: {
+		...buttonReset(),
+		display: 'inline-flex',
+		marginLeft: 10,
+	},
+	upload: {
+		display: 'inline-flex',
+		marginLeft: 10,
+		cursor: 'pointer',
+		transform: 'rotate(180deg)'
+	},
+	input: {
+		...visuallyHidden()
+	}
+})
 
-const DictionaryEditor = ({ dictionary }: DictionaryEditorProps) => {
+const DictionaryEditor = () => {
+
+	const { dictionary, editDictionary } = useSelectedDictionaryStore();
+	const classes = useStyle();
 
 	const dictionaryInputConfig = useInput({
 		initialValue: dictionary?.spec.data,
 		id: 'dictionary-editor',
 		label: dictionary?.name,
+		validate: value => isXMLValid(value)
 	});
 
-	return <ConfigEditor configInput={dictionaryInputConfig} />;
+	const prevValue = usePrevious(dictionaryInputConfig.value);
+
+	const uploadDictionary = async (e: React.ChangeEvent<HTMLInputElement>) => {
+		if (e.target.files) {
+			const file = e.target.files[0];
+			const data = await file.text();
+			dictionaryInputConfig.setValue(data);
+		}
+	}
+
+	const downloadDictionary = () => {
+		if (dictionary) {
+			downloadFile(dictionaryInputConfig.value, dictionary.name, 'text/xml');
+		}
+	}
+
+	return (
+		<>
+			<ConfigEditor configInput={dictionaryInputConfig} />
+			<div className={classes.controls}>
+				<button
+					className={classes.applyChanges}
+					disabled={prevValue === dictionaryInputConfig.value}
+					onClick={() => {
+						if (!dictionaryInputConfig.isValid) {
```

Comment:
isnt it simplier and clearer if we just check if it's valid and then we can edit?